bayesian_changepoint_detection_experiment/01_change_point_detection.py

```diff
@@ -182,79 +244,80 @@
-        title="Changepoint Prediction for Token ('copyright')",
-    )
-)
+# This section uses the CUSUM algorithm to determine changepoint events from time series. This algorithm uses a threshold value to determine cutoffs for a changepoint event. To figure out this value I'm using a cutoff of 2 standard deviations from the mean. The mean is calculated after filtering out one outlier that contains a value of 5000 when the other values are 700 and less. This allows for more potential matches to be found.
 
-(
-    p9.ggplot(change_point_df >> ply.query("tok=='pandemic'"))
-    + p9.aes(x="year_pair", y="changepoint_prob", group=0)
-    + p9.geom_point()
-    + p9.geom_line()
-    + p9.coord_flip()
-    + p9.theme_seaborn("white")
-    + p9.labs(
-        x="Year Shift",
-        y="Probability of Changepoint",
-        title="Changepoint Prediction for Token ('pandemic')",
-    )
+cutoff_df = (
+    change_metric_df
+    >> ply.query("change_metric_ratio < 1000")
+    >> ply.select("change_metric_ratio")
+    >> ply.call(".describe")
 )
+cutoff_df
+
+threshold = cutoff_df.loc["mean"] * cutoff_df.loc["std"] * 2
+threshold[0]
+
+if not Path("output/cusum_changepoint_abstracts.tsv").exists():
+    change_point_results = []
+    for tok, tok_series_df in tqdm.tqdm(change_metric_df.groupby("tok")):
 
-(
-    p9.ggplot(change_point_df >> ply.query("tok=='the'"))
-    + p9.aes(x="year_pair", y="changepoint_prob", group=0)
-    + p9.geom_point()
-    + p9.geom_line()
-    + p9.coord_flip()
-    + p9.theme_seaborn("white")
-    + p9.labs(
-        x="Year Shift",
-        y="Probability of Changepoint",
-        title="Changepoint Prediction for Token ('the')",
+        change_metric_ratio = tok_series_df >> ply.pull("change_metric_ratio")
+        year_series = tok_series_df >> ply.pull("year_pair")
+
+        alarm, start, end, amp = detect_cusum(
+            change_metric_ratio,
+            threshold=threshold[0],
+            drift=0.5,
+            ending=True,
+            show=False,
+        )
+
+        for values in zip(alarm, start, end, amp):
+
+            change_point_results.append(
+                {
+                    "tok": tok,
+                    "changepoint_idx": year_series[values[0]],
+                    "start_idx": year_series[values[1]],
+                    "end_idx": year_series[values[2]],
+                    "value": values[3],
+                }
+            )
+
+if not Path("output/cusum_changepoint_abstracts.tsv").exists():
+    change_point_df = pd.DataFrame.from_records(change_point_results)
+    change_point_df >> ply.call(
+        ".to_csv", "output/cusum_changepoint_abstracts.tsv", sep="\t", index=False
     )
-    + p9.scale_y_continuous(limits=[0, 1])
-)
+else:
+    change_point_df = pd.read_csv("output/cusum_changepoint_abstracts.tsv", sep="\t")
+print((change_point_df >> ply.select("tok") >> ply.distinct()).shape)
+change_point_df
+
+(change_point_df >> ply.arrange("-value") >> ply.slice_rows(30))
```

Comment:
>From your takeaways, it seems that changepoint refers to when a word changes meaning. I'm not sure how acquired the different meanings? Is this from the embedding - so words had a high weight associated with "pandemic" used to be things like "influenza" and "H1N1" but now the words most associated with "pandemic" are "covid-19"?

Ah good questions. Words "change meaning" naturally as we continuously write papers and other forms of literature. Here change meaning is synonymous to change associations. Pandemic doesn't change its meaning in regards to a disease taking over the world, but what does change is the kind of disease that is prevalent at that time period. For example, pandemic was associated with H1N1 because a lot of papers were talking about H1N1. RECENTLY people and scientists started to talk more about covid-19. This event is considered a "changepoint". 

Now on the form of detecting these changes it isn't the embeddings but the combined ratios of cosine distances  between two periods (2019-2020/2018-2019)  and frequency changes between two periods (2020/2019).  Higher values equate to higher chances of a "change".

Reply:
>I think pandemic is a good positive control, is there a word that you expect doesn't change as a negative/baseline?

Stopwords such as "the", "and", "a" etc. These words are everywhere despite the time period, so they shouldn't have a change or be detected as having a change.